system/src/system_connection_manager.cpp

```diff
@@ -0,0 +1,491 @@
+            return SYSTEM_ERROR_BAD_DATA;
+        }
+
+        if (pfds[i].revents & POLLIN) {
+            CHECK(receiveTestPacket(connection));
+        }
+        if (pfds[i].revents & POLLOUT) {
+            CHECK(sendTestPacket(connection));
+        }
+    }
+    return 0;
+};
+
+void ConnectionTester::cleanupSockets(bool recalculateMetrics) {
+    for (auto& i: metrics_) {
+
+        if (recalculateMetrics && i.rxPacketCount > 0) {
+            i.avgPacketRoundTripTime = (i.totalPacketWaitMillis / i.rxPacketCount);
+
+            LOG(INFO,"%s: %d/%d packets %d/%d bytes received, avg rtt: %d", 
+                netifToName(i.interface), 
+                i.rxPacketCount,
+                i.txPacketCount, 
+                i.rxBytes,
+                i.txBytes,
+                i.avgPacketRoundTripTime);
+        }
+
+        sock_close(i.socketDescriptor);
+        i.socketDescriptor = -1;
+        i.txPacketCount = 0;
+        i.rxPacketCount = 0;
+        i.totalPacketWaitMillis = 0;
+
+        if (i.txBuffer) {
+            free(i.txBuffer);
+        }
+        if (i.rxBuffer) {
+            free(i.rxBuffer);
+        }
+    }
+
+    // Sort list by packet latency in ascending order, ie fastest to slowest
+    std::sort(metrics_.begin(), metrics_.end(), [](const ConnectionMetrics& dg1, const ConnectionMetrics& dg2) {
+        return (dg1.avgPacketRoundTripTime < dg2.avgPacketRoundTripTime); 
+    });
+};
+
+
+// GOAL: To maintain a list of which network interface is "best" at any given time
+// 1) Retrieve the server hostname and port. Resolve the hostname to an addrinfo list (ie IP addresses of server)
+// 2) Create a socket for each network interface to test. Bind this socket to the specific interface. Connect the socket
+// 3) Add these created+connected sockets to a pollfd structure. Allocate buffers for the reachability test messages.
+// 4) Poll all the sockets. Polling sends a reachability test message and waits for the response. The test continues for the test duration
+// 5) After polling completes, free the allocated buffers, reset diagnostics and calculate updated metrics. 
+int ConnectionTester::testConnections() {
+    // Step 1: resolve server hostname to IP address
+    struct addrinfo* info = nullptr;
+    struct addrinfo hints = {};
+    hints.ai_flags = AI_NUMERICSERV | AI_ADDRCONFIG;
+    hints.ai_protocol = IPPROTO_UDP;
+    hints.ai_socktype = SOCK_DGRAM;
+
+    char tmphost[128] = {}; // TODO: better size ie sizeof(address->domain)
+    char tmpserv[8] = {};
+    uint16_t tmpport = 0;
+
+    int r = SYSTEM_ERROR_NETWORK;
+    bool testSuccessful = false;
+
+    getCloudHostnameAndPort(&tmpport, tmphost, sizeof(tmphost));
+    snprintf(tmpserv, sizeof(tmpserv), "%u", tmpport);
+    LOG(TRACE, "Resolving %s#%s", tmphost, tmpserv);
+    // TODO: get addrinfo/server IP from DNS lookup using the specific interface? 
+    r = netdb_getaddrinfo(tmphost, tmpserv, &hints, &info); 
```

Comment:
This is becoming a problem when different interfaces DNS servers resolve the device service hostname to different IPs and we try to use the same IP to try to connect with all the interfaces.
IE Some situations im seeing are wifi+ethernet sockets will connect but cellular doesnt and vice versa.

Reply:
For example
Cellular + WIFI DNS servers:
```
[2023-11-28 11:45:14.380] 0000204617 [app] INFO: DNS IP 8.8.4.4
[2023-11-28 11:45:14.380] 0000204621 [app] INFO: DNS IP 8.8.8.8
[2023-11-28 11:45:14.385] 0000204624 [app] INFO: DNS IP 192.168.1.1
```
Server IP Resolves to `3.222.143.118#5684`

Ethernet + Cellular + WIFI DNS servers:
```
[2023-11-28 11:45:51.616] 0000241854 [app] INFO: DNS IP 192.168.0.1
[2023-11-28 11:45:51.622] 0000241858 [app] INFO: DNS IP 8.8.4.4
[2023-11-28 11:45:51.622] 0000241861 [app] INFO: DNS IP 8.8.8.8
[2023-11-28 11:45:51.627] 0000241865 [app] INFO: DNS IP 192.168.1.1
```

Server IP resolves to a different address `3.215.122.234#5684`
For whatever reason, cellular sometimes cannot reach this IP, or the traffic is much slower than with its own resolved IP